Boost rewrite breaks build

Consider the following meson.build file
```
project('test_link', 'cpp', default_options : ['cpp_std=c++17'])

boost_dep = dependency('boost', modules : ['system'])

sources = files(
        'main.cpp'
)

a_out = executable('a.out', sources, dependencies : [boost_dep])
```
(the file `main.cpp` can just contain `int main(){}`.)

This no longer builds.  The behavior was introduced with 08224dafcba1b694fb624553e7d84deb565aae22.

The problem is that instead of `-lboost_system`, the file `/usr/lib/libboost_system.so.1.72.0` is passed as such to the linker:
```
[2/2] c++  -o a.out 'a.out@exe/main.cpp.o' -Wl,--as-needed -Wl,--no-undefined -Wl,--start-group /usr/lib/libboost_system.so.1.72.0 -Wl,--end-group
FAILED: a.out 
c++  -o a.out 'a.out@exe/main.cpp.o' -Wl,--as-needed -Wl,--no-undefined -Wl,--start-group /usr/lib/libboost_system.so.1.72.0 -Wl,--end-group
/usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: error: /usr/lib/libboost_system.so.1.72.0: incompatible target
collect2: error: ld returned 1 exit status
```

(To keep the linker from dropping the shared library altogether, I changed `main.cpp` to
```cpp
#include <boost/filesystem.hpp>

int main()
{
        return boost::filesystem::path("/path/to/something").filename().size() != 0;
}
```
and added `filesystem` as a boost dependency to `meson.build`.)

```sh
fiesh@wsbox /tmp/working % ldd a.out
        linux-vdso.so.1 (0x00007fffbf9f7000)
        libboost_filesystem.so.1.72.0 => /usr/lib64/libboost_filesystem.so.1.72.0 (0x00007f5305d7e000)
        libstdc++.so.6 => /usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/libstdc++.so.6 (0x00007f5305b05000)
        libgcc_s.so.1 => /usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/libgcc_s.so.1 (0x00007f5305aec000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f5305933000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f5305929000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f5305909000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f53057d1000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f5305df8000)
```

I suppose by "compiling by hand" you mean the following?
```
fiesh@wsbox /tmp/notworking % c++  -o a.out 'a.out@exe/main.cpp.o' /usr/lib/libboost_system.so.1.72.0 /usr/lib/libboost_filesystem.so.1.72.0
/usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: error: /usr/lib/libboost_system.so.1.72.0: incompatible target
/usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: error: /usr/lib/libboost_filesystem.so.1.72.0: incompatible target
../t/main.cpp:5: error: undefined reference to 'boost::filesystem::path::filename() const'
collect2: error: ld returned 1 exit status
```

Unlike with static libraries, I don't think it is possible to specify a shared library by its name, one has to use the `-llibname` mechanism.

Running `find / -name "*boost_system*" -type f` is a bit difficult since it's a rather big server.  But I suppose this should suffice:
```sh
% find /usr -name "*boost_system*" -type f 2>/dev/null
/usr/lib/libboost_system.so.1.72.0
/usr/lib/debug/usr/lib64/libboost_system.so.1.72.0.debug
/usr/lib/debug/usr/lib/libboost_system.so.1.72.0.debug
/usr/lib/libboost_system.a
/usr/share/aclocal/ax_boost_system.m4
/usr/lib64/libboost_system.so.1.72.0
/usr/lib64/libboost_system.a
```

--- a/mesonbuild/dependencies/boost.py
+++ b/mesonbuild/dependencies/boost.py
@@ -412,10 +412,19 @@ class BoostDependency(ExternalDependency):
                 break
 
     def run_check(self, inc_dirs: T.List[BoostIncludeDir], lib_dirs: T.List[Path]) -> bool:
+        mlog.debug('  - potential library dirs: {}'.format([x.as_posix() for x in lib_dirs]))
+        mlog.debug('  - potential include dirs: {}'.format([x.path.as_posix() for x in inc_dirs]))
+
         #   2. Find all boost libraries
         libs = []  # type: T.List[BoostLibraryFile]
         for i in lib_dirs:
-            libs += self.detect_libraries(i)
+            libs = self.detect_libraries(i)
+            if libs:
+                mlog.debug('  - found boost library dir: {}'.format(i))
+                # mlog.debug('  - raw library list:')
+                # for j in libs:
+                #     mlog.debug('    - {}'.format(j))
+                break
         libs = sorted(set(libs))
 
         modules = ['boost_' + x for x in self.modules]
@@ -423,9 +432,6 @@ class BoostDependency(ExternalDependency):
             mlog.debug('  - found boost {} include dir: {}'.format(inc.version, inc.path))
             f_libs = self.filter_libraries(libs, inc.version_lib)
 
-            # mlog.debug('  - raw library list:')
-            # for j in libs:
-            #     mlog.debug('    - {}'.format(j))
             mlog.debug('  - filtered library list:')
             for j in f_libs:
                 mlog.debug('    - {}'.format(j))
@@ -500,6 +506,19 @@ class BoostDependency(ExternalDependency):
         return [self._include_dir_from_version_header(x) for x in candidates]
 
     def detect_lib_dirs(self, root: Path) -> T.List[Path]:
+        # First check the system include paths. Only consider those within the
+        # given root path
+        system_dirs_t = self.clib_compiler.get_library_dirs(self.env)
+        system_dirs = [Path(x) for x in system_dirs_t]
+        system_dirs = [x.resolve() for x in system_dirs if x.exists()]
+        system_dirs = [x for x in system_dirs if mesonlib.path_is_in_root(x, root)]
+        system_dirs = list(mesonlib.OrderedSet(system_dirs))
+
+        if system_dirs:
+            return system_dirs
+
+        # No system include paths were found --> fall back to manually looking
+        # for library dirs in root
         dirs = []     # type: T.List[Path]
         subdirs = []  # type: T.List[Path]
         for i in root.iterdir():
@@ -511,7 +530,25 @@ class BoostDependency(ExternalDependency):
             for j in i.iterdir():
                 if j.is_dir() and j.name.endswith('-linux-gnu'):
                     subdirs += [j]
-        return dirs + subdirs
+
+        # Filter out paths that don't match the target arch to avoid finding
+        # the wrong libraries. See https://github.com/mesonbuild/meson/issues/7110
+        if not self.arch:
+            return dirs + subdirs
+
+        arch_list_32 = ['32', 'i386']
+        arch_list_64 = ['64']
+
+        raw_list = dirs + subdirs
+        no_arch = [x for x in raw_list if not any([y in x.name for y in arch_list_32 + arch_list_64])]
+
+        matching_arch = []  # type: T.List[Path]
+        if '32' in self.arch:
+            matching_arch = [x for x in raw_list if any([y in x.name for y in arch_list_32])]
+        elif '64' in self.arch:
+            matching_arch = [x for x in raw_list if any([y in x.name for y in arch_list_64])]
+
+        return sorted(matching_arch) + sorted(no_arch)
 
     def filter_libraries(self, libs: T.List[BoostLibraryFile], lib_vers: str) -> T.List[BoostLibraryFile]:
         # MSVC is very picky with the library tags

--- a/mesonbuild/mesonlib.py
+++ b/mesonbuild/mesonlib.py
@@ -1533,6 +1533,16 @@ def relpath(path: str, start: str) -> str:
     except (TypeError, ValueError):
         return path
 
+def path_is_in_root(path: Path, root: Path, resolve: bool = False) -> bool:
+    # Check wheter a path is within the root directory root
+    try:
+        if resolve:
+            path.resolve().relative_to(root.resolve())
+        else:
+            path.relative_to(root)
+    except ValueError:
+        return False
+    return True
 
 class LibType(Enum):
 

